### PR TITLE
Prevent login! failure blocks from masking InvalidCredentials

### PR DESCRIPTION
**Pull request by willnet**
This PR fixes a bug that I found while trying to remove the mock in #438. 

login! is documented to raise Sorcery::InvalidCredentials when authentication fails. When authenticate yields a failure reason, forwarding the caller's block to login lets controller code in the failure branch run before login! can raise, so errors such as a missing template can replace InvalidCredentials.

Call login without forwarding the block, then yield to the caller only after a successful authentication. This keeps login's failure-reason block behavior intact while preserving login!'s raise-on-failure contract.

Please ensure your pull request includes the following:

- [x] Description of changes
- [x] Changes have related RSpec tests that ensure functionality does not break

### CODE REVIEW

File: lib/sorcery/controller.rb [hunk 1]
```diff
@@ -59,12 +59,12 @@ def login(*credentials)
...
 
-      def login!(...)
-        user = login(...)
+      def login!(*credentials)
+        user = login(*credentials)
 
         raise Sorcery::InvalidCredentials if user.nil?
 
-        user
+        block_given? ? yield(user, nil) : user
```
[the remark this answers is not in the document]
Comment (willnet) [reply lib/sorcery/controller.rb]: > Please add/adjust a spec for #login! with block where User.authenticate yields (user, :invalid_password) and assert that the request raises Sorcery::InvalidCredentials

The spec updated in this PR is the one you're referring to.

[the remark this answers is not in the document]
Comment (brendon) [reply lib/sorcery/controller.rb]: I think here there is still an omission. `user` was always returned previously, now it's only returned if a block wasn't given. Someone might be using a block and then also persisting the return value of `login!`.

File: spec/controllers/sorcery_controller_spec.rb [hunk 2]
```diff
@@ -107,7 +107,7 @@
 
       context 'when fails' do
         before do
-          allow(User).to receive(:authenticate).with('bla@example.com', 'opensesame!').and_return(nil)
+          allow(User).to receive(:authenticate).with('bla@example.com', 'opensesame!').and_yield(user, :invalid_password)
```
Comment (brendon) [inline spec/controllers/sorcery_controller_spec.rb]: It doesn't really matter, but this would be more correct if it yielded `nil` as the second parameter since that's what the real method does. You could probably add `.and_return(user)` to the end also. None of these paths are even followed in the failure case though. You could potentially have neither `and_yield` and `and_return` and it would be fine?

You could add some more tests that test that when you rescue the error, that the user is yielded along with a `nil` second value, and that the user is returned after the block is executed? And a test where there is no block passed.

**Review by brendon — CHANGES_REQUESTED** [review]
